### PR TITLE
Template of Training Pipeline

### PR DESCRIPTION
**Pull request by dupeljan**

### CODE REVIEW

File: train_od_ex.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        plt.rcParams['figure.figsize'] = [14, 7]
+
+        for idx, train_image_np in enumerate(self.train_images_np):
+            plt.subplot(2, 3, idx+1)
+            plt.imshow(train_image_np)
+        plt.show()
+
+
+    def get_gt_boxes(self):
+        self.gt_boxes = [
```
Comment (dupeljan) [inline train_od_ex.py:73]: Potential place to put annotations for our custom dataset

File: train_od_ex.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+from utils import load_image_into_numpy_array
+from utils import plot_detections
+
+OD_MODELS_PATH = '.'
+os.chdir(OD_MODELS_PATH)
+# For interactive plot in PyCharm
+matplotlib.use('module://backend_interagg')
+
+
+class RubberDucksTraining:
```
Comment (kprokofi) [inline train_od_ex.py]: Why do we need DuckTraining at all? How this related to our project?

**Comment by kprokofi** [comment]
@dupeljan Please, consider aligning training with my proposal data loader from kp/data branch.  It's more appropriate to build a dataset separately. The proposal function will return ready-to-go for loop TensorFlow datasets

**Comment by dupeljan** [comment]
@kprokofi it's the simplest use case of standard tensorwlow od model training. Now it's your turn to allingn chosen dataset to this pipeline (it's not a big deal)

File: train_od_ex.py (hunk 2 above)
Comment (kprokofi) [reply train_od_ex.py]: Please, change names and structure more related to our project w/o some rubber ducks

File: train_od_ex.py (hunk 1 above)
Comment (kprokofi) [reply train_od_ex.py:73]: ok

File: train_od_ex.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            tf.config.experimental.set_memory_growth(device, True)
+        self.gt_boxes = []
+        self.load_dataset()
+        self.get_gt_boxes()
+        self.preprocess_data()
+        self.create_model_from_checkpoint()
+        self.finetune()
+        self.test()
+
+    def load_dataset(self):
```
Comment (kprokofi) [inline train_od_ex.py:47]: I will change behavior of this function to load tf dataset. Will I? Anyway, I have a question. For what reason, we need to load images in numpy arrays?

Comment (kprokofi) [reply train_od_ex.py:47]: If the only reason for that is pyplot visualization I suggest removing it and move visualization to another script if you need this because I created a function to load data exactly in tf representation

File: train_od_ex.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+            self.train_image_tensors.append(tf.expand_dims(tf.convert_to_tensor(
+                train_image_np, dtype=tf.float32), axis=0))
+            self.gt_box_tensors.append(tf.convert_to_tensor(gt_box_np, dtype=tf.float32))
+            zero_indexed_groundtruth_classes = tf.convert_to_tensor(
+                np.ones(shape=[gt_box_np.shape[0]], dtype=np.int32) - self.label_id_offset)
+            self.gt_classes_one_hot_tensors.append(tf.one_hot(
+                zero_indexed_groundtruth_classes, self.num_classes))
+        print('Done prepping data.')
+
+    def visualize_np(self):
```
Comment (kprokofi) [inline train_od_ex.py:110]: hmm, do we have two visualization functions?

File: train_od_ex.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+
+        # Load pipeline config and build a detection model.
+        #
+        # Since we are working off of a COCO architecture which predicts 90
+        # class slots by default, we override the `num_classes` field here to be just
+        # one (for our new rubber ducky class).
+        configs = config_util.get_configs_from_pipeline_file(pipeline_config)
+        model_config = configs['model']
+        model_config.ssd.num_classes = num_classes
+        model_config.ssd.freeze_batchnorm = True
```
Comment (kprokofi) [inline train_od_ex.py:139]: Why do we freezing batch normalization for fine-tuning?

File: train_od_ex.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        self.detection_model = model_builder.build(
+            model_config=model_config, is_training=True)
+
+        # Set up object-based checkpoint restore --- RetinaNet has two prediction
+        # `heads` --- one for classification, the other for box regression.  We will
+        # restore the box regression head but initialize the classification head
+        # from scratch (we show the omission below by commenting out the line that
+        # we would add if we wanted to restore both heads)
+        fake_box_predictor = tf.compat.v2.train.Checkpoint(
```
Comment (kprokofi) [inline train_od_ex.py:148]: Can you explain why do we need some fake heads and stuff below? Actually, all we need to do is to load weights for the backbone and initialize a new head for classification. Don't we? Or these weird and non-intuitive steps are required by TensorFlow?

File: train_od_ex.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        _ = self.detection_model.postprocess(prediction_dict, shapes)
+        print('Weights restored!')
+
+    def finetune(self):
+        tf.keras.backend.set_learning_phase(True)
+
+        # These parameters can be tuned; since our training set has 5 images
+        # it doesn't make sense to have a much larger batch size, though we could
+        # fit more examples in memory if we wanted to.
+        batch_size = 4
```
Comment (kprokofi) [inline train_od_ex.py]: hardcoded numbers. Please, do it in a more explicit way. As a last resort, make global variables.

File: train_od_ex.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        learning_rate = 0.01
+        num_batches = 100
+
+        # Select variables in top layers to fine-tune.
+        trainable_variables = self.detection_model.trainable_variables
+        to_fine_tune = []
+        prefixes_to_train = [
+            'WeightSharedConvolutionalBoxPredictor/WeightSharedConvolutionalBoxHead',
+            'WeightSharedConvolutionalBoxPredictor/WeightSharedConvolutionalClassHead']
+        for var in trainable_variables:
```
Comment (kprokofi) [inline train_od_ex.py:182]: Does it mean that we fine-tune only heads?

File: train_od_ex.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+            _box_prediction_head=self.detection_model._box_predictor._box_prediction_head,
+        )
+        fake_model = tf.compat.v2.train.Checkpoint(
+            _feature_extractor=self.detection_model._feature_extractor,
+            _box_predictor=fake_box_predictor)
+        ckpt = tf.compat.v2.train.Checkpoint(model=fake_model)
+        ckpt.restore(checkpoint_path).expect_partial()
+
+        # Run model through a dummy image so that variables are created
+        image, shapes = self.detection_model.preprocess(tf.zeros([1, 640, 640, 3]))
```
Comment (kprokofi) [inline train_od_ex.py:161]: What does the preprocess function? It's important for me to know

File: utils.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+import matplotlib.pyplot as plt
+
+import numpy as np
+from six import BytesIO
+from PIL import Image
+
+import tensorflow as tf
+from object_detection.utils import visualization_utils as viz_utils
+
+def load_image_into_numpy_array(path):
```
Comment (kprokofi) [inline utils.py:10]: Another function to load data to NumPy?

File: utils.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    Returns:
+      uint8 numpy array with shape (img_height, img_width, 3)
+    """
+    img_data = tf.io.gfile.GFile(path, 'rb').read()
+    image = Image.open(BytesIO(img_data))
+    (im_width, im_height) = image.size
+    return np.array(image.getdata()).reshape(
+        (im_height, im_width, 3)).astype(np.uint8)
+
+def plot_detections(image_np,
```
Comment (kprokofi) [inline utils.py:29]: The third function for visualization? :)

**Review by kprokofi — CHANGES_REQUESTED** [review]

File: train_od_ex.py (hunk 3 above)
Comment (dupeljan) [reply train_od_ex.py:47]: Yeah, it only an example. I'll remove it and when you will add appropriate tf loader instead. I can't test pipeline without this method

File: train_od_ex.py (hunk 4 above)
Comment (dupeljan) [reply train_od_ex.py:110]: It's exist to visualize images in numpy ^)

File: train_od_ex.py (hunk 5 above)
Comment (dupeljan) [reply train_od_ex.py:139]: Suggestion of official tf tutorial. We can ignore it and tune whatever we want

File: train_od_ex.py (hunk 6 above)
Comment (dupeljan) [reply train_od_ex.py:148]: It's weird way to load model from checkpoint without classification head. TF checkpoints is really difficult to load partially thus this hack exist

File: train_od_ex.py (hunk 9 above)
Comment (dupeljan) [reply train_od_ex.py:161]: Please read line https://github.com/kprokofi/animal-recognition-with-voice/pull/1/files#diff-dd1d550d71fd1ed9372fc1021bad76848d71ab8ca4e72ab35c6a14c4ba3a980dR147. It's way to initialize operation shapes

File: train_od_ex.py (hunk 7 above)
Comment (dupeljan) [reply train_od_ex.py]: Ok, I'll make an attributes

File: train_od_ex.py (hunk 8 above)
Comment (dupeljan) [reply train_od_ex.py:182]: I'm not sure, we have to inspect graph. But I think it tunes only head, yeah. We can choose layers whatever we want

File: utils.py (hunk 10 above)
Comment (dupeljan) [reply utils.py:10]: It loads single image

File: utils.py (hunk 11 above)
Comment (dupeljan) [reply utils.py:29]: This function quite useful because it visualize bounding boxes on images

File: train_od_ex.py (hunk 2 above)
Comment (dupeljan) [reply train_od_ex.py]: Rubber ducks are quite good as simplest validation dataset
